### ISSUE TITLE
ColorBys with Missing Tip Values & Not in `colors.tsv` Break Auspice 2.0

### ISSUE THREAD

**Issue by emmahodcroft** [nextstrain/auspice#774]
Running Auspice `v2` branch, with JSONs generated from `augur` `v6` branch using `export v2`. 

The current Enterovirus D68 build as on the Nextstrain website, I've used for lots of testing - all colorbys work great here with all the v2/v6 stuff. 

However, today I tried to add 2 new traits, "source"  and "symptom". The important thing to know about these is that *not all tips have a value for this* (they have missing tip values) and *they have no value in `colors.tsv`*. 

This makes them different from other colorbys (which all work):

Country and region are categorical, have no missing tip values, and are in `colors.tsv`. 🔴✔️✔️
Clade is categorical, has no missing tip values, and is _not_ in `colors.tsv`. 🔴✔️❌ 
Host age is continuous, has missing tip values, and is _not_ in `colors.tsv` 🔵❌❌ 
Age range 1, 2, and 3, are categorical, have missing tip values, and are in `colors.tsv` 🔴❌✔️

Source and symptom are categorical, have missing tip values, and are _not_ in `colors.tsv`  🔴❌❌

_Clade gives the value "Unassigned" to those not in a clade, so does not technically have any missing values_
_(I'm excluding GT and date because these are a bit different and the same rules don't always apply.)_

I can break `age_range3` in the same way by removing it from `colors.tsv`. I can also make it break by including boolean data (true/false, 0/1, etc) that's not in `colors.tsv` if it has any missing values. (It works if it has no missing values.)

So this is the bad combo: categorical or boolean, missing tip values, not in `colors.tsv`.

----

The error occurs when trying to create the pie-chart circles:
![image](https://user-images.githubusercontent.com/14290674/63185370-244b1200-c05a-11e9-8073-aa85124f2ca9.png)
(Line numbers are off because of my debug. Real line is line [97](https://github.com/nextstrain/auspice/blob/998d46e271e4638b3afe35849253d21eb24b2aa8/src/components/map/mapHelpersLatLong.js#L97) in `mapHelptersLatLong.js`)

----

I _believe_ the problem is due to the fact that there are no `undefined` counts, which mean that when counting up the number of traits in a pie-chart, they do not equal to the total number of things in that pie-chart. 

So as an example with `age_range2`, `legendValues`, which is used to map counts of each possible option, has four values, including `undefined`:
![image](https://user-images.githubusercontent.com/14290674/63185809-39747080-c05b-11e9-9067-1aa62f36518e.png)

Then the `legendValueToArcIdx` which holds the counts, has counts for each:
![image](https://user-images.githubusercontent.com/14290674/63185838-498c5000-c05b-11e9-9cc6-4503e20d90c1.png)

If I switch to `age_range3` (when I've broken it), it has no `undefined` in either:
![image](https://user-images.githubusercontent.com/14290674/63185957-86f0dd80-c05b-11e9-995b-2373cb4127f6.png)
(this is just before it breaks)

If I unbreak it by putting it back in `colors.tsv`, it has `undefined`:
![image](https://user-images.githubusercontent.com/14290674/63186032-b1429b00-c05b-11e9-842b-67931a19986b.png)

----

So in short, I believe the problem is that for categorical and boolean colorBys with missing tip values and no entry in `colors.tsv`, for some reason they never are given an `undefined` category. It looks like maybe this should be happening in association with the Legend stuff, since the pie-chart counts depend on `legendValues`. 

The very tricky thing about this problem is that it raises no errors, warnings, etc, util you actually load it up in Auspice and click on the colorby dropdown - unlikely to be caught by tests!

**Comment by emmahodcroft** [comment]
That's as far as I got, but hopefully enough info that it gives a good start to a fix...!

**Comment by emmahodcroft** [comment]
I haven't been able to shed any further light on this unfortunately, but it would be really good to fix! It's holding me back from casual (as part of other work) v2 testing as not being able to actually color-by all my colorbys is not good...

**Comment by jameshadfield** [comment]
Hey @emmahodcroft i'm just diving into this now but can't yet reproduce the bug.

I've modified the auspice provided minimal JSON (in `./examples`) and removed the color-by values for some tips, with no resulting errors. I've attached the modified JSON below, and this is what it looks like:

![image](https://user-images.githubusercontent.com/8350992/63820806-c940e580-c99e-11e9-944a-93d10d7be5a7.png)

I'll try to recreate something in augur as i'm now thinking this is an augur bug. If I can't, i'll get you to create a small test build in augur which reproduces this.

[minimal_v2.json.zip](https://github.com/nextstrain/auspice/files/3548821/minimal_v2.json.zip)

**Comment by jameshadfield** [comment]
@emmahodcroft could you test https://github.com/nextstrain/auspice/pull/780 with your example? That PR fixes it for the JSON I attached earlier. If there are still issues, could you post your JSONs / create a test build in augur which reproduces the problem.

**Comment by emmahodcroft** [comment]
So weirdly I'm struggling to replicate this now, but Richard points out it's on the v2 version online: https://nextstrain-dev.herokuapp.com/enterovirus?c=symptom

I just pulled all the latest from `augur` and `auspice` and re-ran everything (but not yet tried #780 as you suggest). 

Weirdly I am now having trouble setting `geo` to country - it's erroring and revering to region. Will investigate to see if this is related or separate...

**Comment by emmahodcroft** [comment]
Ah! It's working because I added `source` and `symptom` to my `colors.tsv` so it would work for a presentation last week..... Let me undo and try again.

**Comment by emmahodcroft** [comment]
Yes! That fixes it! Thank you!

The `geo` setting issue seems to be on the `augur` side - it's not getting exported. Will look into and report.

**Comment by emmahodcroft** [comment]
Oh - here's a thing though:

Here's how the pie charts show up for a trait with colours defined in `colors.txt`:
(https://nextstrain-dev.herokuapp.com/enterovirus?c=age_range1)
![image](https://user-images.githubusercontent.com/14290674/63860217-03af8000-c9a9-11e9-9883-c0777b4116b9.png)

Here's how it looks with the fix from #780:
![image](https://user-images.githubusercontent.com/14290674/63860287-20e44e80-c9a9-11e9-9094-9ca462cee6ca.png)

I think this is a bit weird myself, and rather prefer the pie to be coloured grey for missing data. Especially when you get to traits where there are very few tips with any information:
![image](https://user-images.githubusercontent.com/14290674/63860369-483b1b80-c9a9-11e9-8239-8dd09e86ed76.png)

**Comment by jameshadfield** [comment]
@emmahodcroft thanks for the reproducible example, that really helped. Closed in ad0c9b498556ea84691950e62af20f06c9635ffc.

For interest's sake: the problem was to do with what values the color scale function considered present in the tree, and it was subtly different if you provided a `scale` vs if not. I've now updated how we deal with missing or non-valid attributes all across the tree. 

P.S. the reason that "unassigned" shows up in the legend is that it's not a value that the [isValueValid()](https://github.com/nextstrain/auspice/blob/v2/src/util/globals.js#L187) considers invalid. You can add it there if you want to not show it in the legend!

![image](https://user-images.githubusercontent.com/8350992/63901699-dff84280-ca59-11e9-8307-1544d7cfa0d3.png)
